File: app/src/main/java/com/yb/uadnd/matchcentre/AppRepository.java
package com.yb.uadnd.matchcentre;

import android.app.Application;
import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.yb.uadnd.matchcentre.model.Commentary;
import com.yb.uadnd.matchcentre.model.Commentary.Data.CommentaryEntry;
import com.yb.uadnd.matchcentre.model.Match;
import com.yb.uadnd.matchcentre.model.MatchFeedApiInterface;
import com.yb.uadnd.matchcentre.model.database.Comment;
import com.yb.uadnd.matchcentre.model.database.MovieDatabase;
import com.yb.uadnd.matchcentre.model.database.MatchInfo;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class AppRepository {

    private static final String BASE_URL = "https://feeds.incrowdsports.com/provider/opta/football/v1/matches/";
    private static AppRepository appRepository;
    private MatchFeedApiInterface mApi;
    private MovieDatabase mDb;
    private SimpleIdlingResource mRes;
    private MyApp mApp;
    private Executor mExecutor = Executors.newSingleThreadExecutor();

    private AppRepository(Application application) {
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
        mApi = retrofit.create(MatchFeedApiInterface.class);
        mApp = (MyApp) application;
        mDb = mApp.getDatabase();
        mRes = MyApp.Companion.getIdlingResource();
    }

    public static AppRepository getInstance(Application application){
        if(appRepository == null){
            appRepository = new AppRepository(application);
        }
        return appRepository;
    }

    public void deleteOldMatchComments(@NotNull String matchId) {
        mExecutor.execute(new Runnable() {
            @Override
            public void run() {
                mDb.getCommentDao().deleteAllMatchComments(Integer.parseInt(matchId));
            }
        });
    }

    public void fetchMatchCommentaryAndCacheInDb(String matchId){
        Call<Commentary> call = mApi.getMatchCommentary(matchId);
        if(mRes != null) mRes.setIdleState(false);
        call.enqueue(new Callback<Commentary>() {
            @Override
            public void onResponse(Call<Commentary> call, Response<Commentary> response) {
                Log.i("url: ",call.request().toString());
                Commentary comm = response.body();
                if(comm != null){
                    Commentary.Data data = comm.getData();
                    if(data != null) {
                        final MatchInfo info = new MatchInfo(data.getId(), Integer.parseInt(matchId),
                                data.getHomeTeamName(), data.getHomeTeamId(), data.getHomeScore(),
                                data.getAwayTeamName(), data.getAwayTeamId(), data.getAwayScore(),
                                data.getCompetitionId(), data.getCompetition());

                        ArrayList<CommentaryEntry> entries = data.getCommentaryEntries();
                        mExecutor.execute(new Runnable() {
                            @Override
                            public void run() {
                                mDb.getMatchInfoDao().insertMatchInfo(info);
                                if(entries != null) {
                                    for (CommentaryEntry entry : entries) {
                                        mDb.getCommentDao().insertComment(
                                                new Comment(Integer.parseInt(matchId), entry));
                                    }
                                }
                            }
                        });
                    }
                }
            }
            @Override
            public void onFailure(Call<Commentary> call, Throwable t) {

            }
        });
    }

    public LiveData<Match> fetchMatch(String matchId){
        MutableLiveData<Match> matchMutableLiveData = new MutableLiveData<>();
        Call<Match> call = mApi.getMatch(matchId);
        call.enqueue(new Callback<Match>() {
            @Override
            public void onResponse(Call<Match> call, Response<Match> response) {
                Log.i("url: ",call.request().toString());
                Match match = response.body();
                if(match != null){
                    matchMutableLiveData.setValue(match);
                }
            }

            @Override
            public void onFailure(Call<Match> call, Throwable t) {
                Log.i("Retrofit Call Failed", t.getMessage());
            }
        });
        return matchMutableLiveData;
    }

    @NotNull
    public LiveData<List<Comment>> getMatchComments(@NotNull String matchId) {
        return mDb.getCommentDao().getAllMatchComments(Integer.parseInt(matchId));
    }

    @NotNull
    public LiveData<MatchInfo> getMatchInfo(@NotNull String matchId) {
        return mDb.getMatchInfoDao().getMatchInfo(Integer.parseInt(matchId));
    }
}

File: app/src/main/java/com/yb/uadnd/matchcentre/AppRepository.kt
package com.yb.uadnd.matchcentre

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.yb.uadnd.matchcentre.data.Event
import com.yb.uadnd.matchcentre.data.Match
import com.yb.uadnd.matchcentre.data.MatchData
import com.yb.uadnd.matchcentre.data.MatchService
import com.yb.uadnd.matchcentre.data.local.Comment
import com.yb.uadnd.matchcentre.data.local.MatchCentreDatabase
import com.yb.uadnd.matchcentre.data.local.MatchInfo
import io.reactivex.Scheduler
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.addTo
import io.reactivex.rxkotlin.subscribeBy
import timber.log.Timber
import javax.inject.Inject

class AppRepository @Inject constructor(
    private var matchService: MatchService,
    private val db: MatchCentreDatabase,
    private val io: Scheduler,
    private val ui: Scheduler,
    private val idlingResource: SimpleIdlingResource
) {
    private val refreshIntervalSeconds = 3600  // 1 hour
    private val disposables = CompositeDisposable()

    fun getMatchCommentary(newMatchId: Int): LiveData<List<Comment>> {
        var lastRefresh: Long = 0
        db.matchInfoDao.getLastRefreshTime(newMatchId)
            .subscribeOn(io)
            .observeOn(ui)
            .subscribeBy(
                onError = { Timber.e(it)},
                onSuccess = {
                    if (it.isNotEmpty()) lastRefresh = it.first()
                    if(System.currentTimeMillis()/1000 - lastRefresh > refreshIntervalSeconds)
                        refreshCommentary(newMatchId)
                }
            ).addTo(disposables)

        return db.commentDao.getAllMatchComments(newMatchId)
    }

    private fun refreshCommentary(newMatchId: Int) {
        idlingResource.setIdleState(false)
        Timber.d("Fetching comments for match $newMatchId")
        matchService.getMatchCommentary(newMatchId.toString())
            .subscribeOn(io)
            .observeOn(io)
            .subscribeBy(
                onSuccess = { commentary ->
                    idlingResource.setIdleState(true)
                    db.commentDao.deleteAllMatchComments(newMatchId)
                    commentary?.data?.let{
                        val info = MatchInfo.from(data = it)
                        db.matchInfoDao.insertMatchInfo(info)
                            .subscribe()
                        it.commentaryEntries?.let{ entries ->
                            entries.forEach { entry ->
                                db.commentDao.insertComment(Comment(newMatchId, entry))
                            }
                        }
                    }
                },
                onError = {
                    idlingResource.setIdleState(true)
                    Timber.e(it)
                }
            ).addTo(disposables)
    }

    fun fetchMatch(matchId: String): LiveData<Match> {
        idlingResource.setIdleState(false)
        val match = MutableLiveData<Match>()
        matchService.getMatch(matchId)
            .subscribeOn(io)
            .observeOn(ui)
            .subscribeBy(
                onSuccess = {
                    match.value = updateMatchEvents(it)
                    idlingResource.setIdleState(true)
                },
                onError = {
                    idlingResource.setIdleState(true)
                    Timber.e(it)
                }
            ).addTo(disposables)
        return match
    }

    private fun updateMatchEvents(match: Match): Match {
        match.data?.also { data ->
            data.events?.forEach {
                it.updateImageUrl(getEventTeamUrl(data, it))
            }
        }
        return match
    }

    private fun getEventTeamUrl(data: MatchData, event: Event): String? {
        val homeTeamId = data.homeTeam?.id
        val awayTeamId = data.awayTeam?.id
        return when(event.teamId) {
            homeTeamId -> data.homeTeam?.imageUrl
            awayTeamId -> data.awayTeam?.imageUrl
            else -> null
        }
    }

    fun getMatchInfo(matchId: String): LiveData<MatchInfo>{
        return db.matchInfoDao.getMatchInfo(matchId.toInt())
    }

    companion object {
        private var instance: AppRepository? = null

        fun getInstance(
            service: MatchService,
            roomDatabase: MatchCentreDatabase,
            io: Scheduler,
            ui: Scheduler,
            res: SimpleIdlingResource
        ): AppRepository {
            if(instance == null) {
                instance = AppRepository(service, roomDatabase, io, ui, res)
            }
            return instance as AppRepository
        }
    }
}